Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1628 @@
+#align one_hom_class OneHomClass
+
+@[to_additive]
+instance OneHom.oneHomClass : OneHomClass (OneHom M N) M N where
+  coe := OneHom.toFun
+  coe_injective' f g h := by cases f; cases g; congr
+  map_one := OneHom.map_one'
+#align one_hom.one_hom_class OneHom.oneHomClass
+#align zero_hom.zero_hom_class ZeroHom.zeroHomClass
+
+@[simp, to_additive]
+theorem map_one [OneHomClass F M N] (f : F) : f 1 = 1 :=
+  OneHomClass.map_one f
+#align map_one map_one
+#align map_zero map_zero
+
+@[to_additive]
+theorem map_eq_one_iff [OneHomClass F M N] (f : F) (hf : Function.Injective f) {x : M} :
+  f x = 1 ↔ x = 1 := hf.eq_iff' (map_one f)
+#align map_eq_one_iff map_eq_one_iff
+#align map_eq_zero_iff map_eq_zero_iff
+
+@[to_additive]
+theorem map_ne_one_iff {R S F : Type _} [One R] [One S] [OneHomClass F R S] (f : F)
+  (hf : Function.Injective f) {x : R} : f x ≠ 1 ↔ x ≠ 1 := (map_eq_one_iff f hf).not
+#align map_ne_one_iff map_ne_one_iff
+#align map_ne_zero_iff map_ne_zero_iff
+
+@[to_additive]
+theorem ne_one_of_map {R S F : Type _} [One R] [One S] [OneHomClass F R S] {f : F} {x : R}
+  (hx : f x ≠ 1) : x ≠ 1 := ne_of_apply_ne f <| ne_of_ne_of_eq hx (map_one f).symm
+#align ne_one_of_map ne_one_of_map
+#align ne_zero_of_map ne_zero_of_map
+
+/-- Any type that is a `OneHomClass` can be cast into a `OneHom`. -/
+@[to_additive "Any type that is a `ZeroHomClass` can be cast into a `ZeroHom`."]
+instance [OneHomClass F M N] : CoeTC F (OneHom M N) :=
+  ⟨fun f => { toFun := f, map_one' := map_one f }⟩
+
+@[simp, to_additive]
+theorem OneHom.coe_coe [OneHomClass F M N] (f : F) : ((f : OneHom M N) : M → N) = f := rfl
+#align one_hom.coe_coe OneHom.coe_coe
+#align zero_hom.coe_coe ZeroHom.coe_coe
+
+end One
+
+section Mul
+
+variable [Mul M] [Mul N]
+
+/-- `M →ₙ* N` is the type of functions `M → N` that preserve multiplication. The `ₙ` in the notation
+stands for "non-unital" because it is intended to match the notation for `NonUnitalAlgHom` and
+`NonUnitalRingHom`, so a `MulHom` is a non-unital monoid hom.
+
+When possible, instead of parametrizing results over `(f : M →ₙ* N)`,
+you should parametrize over `(F : Type*) [MulHomClass F M N] (f : F)`.
+When you extend this structure, make sure to extend `MulHomClass`.
+-/
+@[to_additive]
+structure MulHom (M : Type _) (N : Type _) [Mul M] [Mul N] where
+  /-- The underlying function -/
+  toFun : M → N
+  /-- The proposition that the function preserves multiplication -/
+  map_mul' : ∀ x y, toFun (x * y) = toFun x * toFun y
+#align mul_hom MulHom
+
+/-- `M →ₙ* N` denotes the type of multiplication-preserving maps from `M` to `N`. -/
+infixr:25 " →ₙ* " => MulHom
+
+/-- `MulHomClass F M N` states that `F` is a type of multiplication-preserving homomorphisms.
+
+You should declare an instance of this typeclass when you extend `MulHom`.
+-/
+-- Porting note: see porting note of `ZeroHomClass`
+@[to_additive]
+class MulHomClass (F : Type _) (M N : outParam (Type _)) [outParam (Mul M)] [outParam (Mul N)]
```

Comment:
```suggestion
class MulHomClass (F : Type _) (M N : Type _) [outParam <| Mul M] [outParam <| Mul N]
```
See [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20timeout).

I am by far no expert, maybe @semorrison understands things better.

The following fails because it tries to synth `MulHomClass (M ≃* N) N (? →ₙ* ?)` first when `M N` are marked as `outParam`. This leads to a timeout as `Pi.instMul` applies to virtually every Pi-type so it expands `(? →ₙ* ?)` even further.
```
import Mathlib.Algebra.Hom.Group
import Mathlib.Logic.Equiv.Basic
import Mathlib.Data.Pi.Algebra

structure MulEquiv (M N : Type _) [Mul M] [Mul N] extends M ≃ N, M →ₙ* N

infixl:25 " ≃* " => MulEquiv

class MulEquivClass (F A B : Type _) [outParam <| Mul A] [outParam <| Mul B] extends EquivLike F A B where
  map_mul : ∀ (f : F) (a b), f (a * b) = f a * f b

instance (priority := 100) [Mul M] [Mul N] [h : MulEquivClass F M N] : MulHomClass F M N :=
  { h with coe := h.coe, coe_injective' := FunLike.coe_injective' }

instance [Mul M] [Mul N] : MulEquivClass (M ≃* N) M N where
  coe f := f.toFun
  inv f := f.invFun
  left_inv f := f.left_inv
  right_inv f := f.right_inv
  coe_injective' f g h₁ h₂ := by
    cases f
    cases g
    congr
    apply Equiv.coe_fn_injective h₁
  map_mul := MulEquiv.map_mul'

variable [Mul M] [Mul N]
#synth MulHomClass (M ≃* N) M N -- this fails
```

Reply:
`(M N : outParam (Type _))` is from mathlib3. From here, `outParam` in `Mul A` and `Mul B` was added because Lean4 couldn't automatically treat instances that depend on `outParam`s as `outParam`. The latter was resolved in issue#1852, but the behaviour of `outParam` still does not mirror Lean3.